### PR TITLE
(dialects): add linalg.tranpose

### LINKED ISSUE #2275: dialects: (linalg) add linalg.transpose
This is a prerequisite for #2272

### PR DESCRIPTION
**Pull request by kayode-gif**
resolves #2275

### CODE REVIEW

**Comment by kayode-gif** [comment]
facing pyright issues based on using `DenseArrayBase.from_list(i64, permutation)`,

File: tests/filecheck/mlir-conversion/with-mlir/dialects/linalg/invalid_ops.mlir [hunk 1]
```diff
@@ -9,4 +9,38 @@ builtin.module {
 
 }
 
+// -----
+
+builtin.module {
+  %0, %1 = "test.op"() : () -> (tensor<16x64xf32>, tensor<64x16x1xf32>)
+
+  // CHECK: Operation does not verify: Input rank (2) does not match init rank (3)
```
Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/linalg/invalid_ops.mlir:17]:
```suggestion
  // CHECK: Operation does not verify: Input rank (2) does not match output rank (3)
```

File: xdsl/dialects/linalg.py [hunk 2]
```diff
@@ -490,6 +496,112 @@ def __init__(
...
+        parser.parse_characters("outs")
+        parser.parse_punctuation("(")
+        init = parser.parse_operand()
+        parser.parse_punctuation(":")
+        # result type the same type as init
+        result = parser.parse_type()
+        parser.parse_punctuation(")")
+        parser.parse_keyword("permutation")
+        parser.parse_punctuation("=")
+        permutation = parser.parse_comma_separated_list(
```
Comment (superlopuh) [inline xdsl/dialects/linalg.py:593]: out of curiosity, what is the parsed type when this line is just `permutation = parser.parse_attribute()`? I'd expect it to parse an array of i64

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/linalg.py (hunk 2 above)
Comment (kayode-gif) [reply xdsl/dialects/linalg.py:593]: ` Operation does not verify: [1 : i64, 0 : i64] should be of base attribute array`
it parses an array yes, but it needs to be a dense array

Comment (superlopuh) [reply xdsl/dialects/linalg.py:593]: kk, thank you

**Review by compor — APPROVED** [review]

**Pull request merged**